test/replication-luatest/linearizable_read_test.lua

```diff
@@ -0,0 +1,273 @@
+    for i = 1, num_instances do
+        table.insert(t, server.build_instance_uri('server_' .. i .. '_proxy'))
+    end
+    return t
+end
+
+local num_servers = 3
+
+g.before_all(function()
+    g.cluster = cluster:new({})
+    g.servers = {}
+    local box_cfg = {
+        replication = build_replication(num_servers),
+        replication_timeout = 0.1,
+        txn_isolation = 'read-confirmed',
+        memtx_use_mvcc_engine = true,
+    }
+    g.servers[1] = g.cluster:build_and_add_server({
+        alias = 'server_1',
+        box_cfg = box_cfg,
+    })
+    -- Servers 2 and 3 are interconnected without a proxy.
+    for i = 2, num_servers do
+        box_cfg.replication[i] = server.build_instance_uri('server_' .. i)
+    end
+    for i = 2, num_servers do
+        g.servers[i] = g.cluster:build_and_add_server({
+            alias = 'server_' .. i,
+            box_cfg = box_cfg,
+        })
+    end
+    g.proxies = {}
+    for i = 1, num_servers do
+        g.proxies[i] = proxy:new({
+            client_socket_path = server.build_instance_uri('server_' .. i .. '_proxy'),
+            server_socket_path = server.build_instance_uri('server_' .. i),
+        })
+        g.proxies[i]:start({force = true})
+    end
+    g.cluster:start()
+    g.cluster:wait_fullmesh()
+    g.servers[1]:exec(function()
+        box.schema.space.create('sync', {is_sync=true})
+        box.space.sync:create_index('pk')
+    end)
+    g.servers[2]:wait_vclock_of(g.servers[1])
+    g.servers[3]:wait_vclock_of(g.servers[1])
+end)
+
+g.after_all(function()
+    g.cluster:drop()
+end)
+
+g.test_wait_others = function(g)
+    g.proxies[1]:pause()
+    local fid = g.servers[1]:exec(function()
+        local t = require('luatest')
+        local f = require('fiber').new(function()
+            return pcall(box.begin, {linearizable = true})
+        end)
+        f:set_joinable(true)
+        t.assert_equals(f:status(), 'suspended', 'begin() waits for replicas')
+        return f:id()
+    end)
+    g.proxies[1]:resume()
+    local ok = g.servers[1]:exec(function(fid)
+        local fiber = require('fiber')
+        local _, ok, err =  fiber.join(fiber.find(fid))
+        return ok, err
+    end, {fid})
+    t.assert(ok, 'begin() succeeds after replica connections are established')
+end
+
+g.test_timeout = function(g)
+    g.servers[1]:exec(function()
+        local t = require('luatest')
+        box.error.injection.set('ERRINJ_RELAY_FROM_TX_DELAY', true)
+        local ok, err = pcall(box.begin, {linearizable = true, timeout = 0.01})
+        box.error.injection.set('ERRINJ_RELAY_FROM_TX_DELAY', false)
+        t.assert(not ok, 'Error when relay is unresponsive')
+        t.assert_equals(err.type, 'TimedOut',
+                        'Timeout when relay is unresponsive')
+    end)
+end
+
+g.before_test('test_no_dirty_reads', function(g)
+    -- Make cluster size 5. Having quorum 3 out of 5 will make the instance
+    -- poll 5 - 3 + 1 == all 3 of the "real" nodes.
```

Comment:
You could also keep cluster size 3 and block server1<->server3 connection via the proxy. Then server1 would have to wait for server2.

Reply:
Yes, but then I'd have to use 2 separate proxies between server1<>server2 and server1<>server3.

Currently proxies[3] blocks server3 relays and server1 appliers.
And proxies[1] blocks all server1 relays.

So it seems easier for me to make the node poll every other node.

Also, I could set `replication_synchro_quorum` to 1 on server1. It would make it query 3 - 1 + 1 = 3. Everyone.
But I decided to use a less hacky way to set `replication_connect_quorum`. Who knows, maybe one day quorum will always be the majority.

Anyway, I'm fine with setting quorum to 1 on server1, if you want me to

Reply:
Oh, somewhy I had an impression that your proxy schema was `server1 <-> proxyN <-> serverN` with N >= 2. But if server1 has only one front proxy, then lets leave the cluster size 5.